controlplane/kubeadm/internal/control_plane.go

```diff
@@ -208,17 +203,123 @@ func (c *ControlPlane) NewMachine(infraRef, bootstrapRef *corev1.ObjectReference
+		Filter(machinefilters.Not(machinefilters.IsFailed))
+
+	if c.Machines.Len() < minimumClusterSizeForRemediation {
+		return machines
+	}
+	return machines.Filter(machinefilters.Not(machinefilters.NeedsRemediation))
+}
+
+// NextMachineForScaleDown returns the next machine that should be selected for removal.
+// It will first recommend the oldest unhealthy machine, if there are any.
+// It then will recommend the oldest outdated machine in the failure domain with the most outdated machines, if there are any.
+// Lastly it will recommend the oldest machine in the largest failure domain.
+func (c *ControlPlane) NextMachineForScaleDown() *clusterv1.Machine {
+	if c.UnhealthyMachines().Any() {
+		machine := c.UnhealthyMachines().Oldest()
+		c.Logger.Info("Scaling down unhealthy machine", "machine", machine.Name)
+		return machine
+	}
+	if c.MachinesNeedingRollout().Any() {
+		machine := c.MachineInFailureDomainWithMostMachines(c.MachinesNeedingRollout())
+		c.Logger.Info("Scaling down outdated machine", "machine", machine.Name)
+		return machine
+	}
+	machine := c.MachineInFailureDomainWithMostMachines(c.Machines)
+	c.Logger.Info("Scaling down unneeded machine", "machine", machine.Name)
+	return machine
+}
+
+// ReadyMachines returns machines that are not provisioning or failed.  In the
+// case of 3 node or larger clusters, it excludes unhealthy machines.
+func (c *ControlPlane) ReadyMachines() FilterableMachineCollection {
+	machines := c.Machines.
+		Filter(machinefilters.Not(machinefilters.IsProvisioning)).
+		Filter(machinefilters.Not(machinefilters.IsFailed))
+
+	if c.Machines.Len() < minimumClusterSizeForRemediation {
+		return machines
+	}
+	return machines.Filter(machinefilters.Not(machinefilters.NeedsRemediation))
+}
+
+// NeedsInitialization returns whether the control plane has yet to create the
+// first machine.  Returns false in the case that zero replicas are desired.
+func (c *ControlPlane) NeedsInitialization() bool {
+	return c.Machines.Len() == 0 && *c.KCP.Spec.Replicas > 0
+}
+
+// NeedsScaleDown returns whether the control plane needs to remove a machine
+// from the cluster.  Scale down can be caused by the presence of unhealthy
+// machines or when there are more machines than desired replicas.
+func (c *ControlPlane) NeedsScaleDown() bool {
+	if c.UnhealthyMachines().Any() {
+		c.Logger.Info("Scale down is required because unhealthy machines were found", "machines", c.UnhealthyMachines().Names())
+		return true
+	}
+	if c.Machines.Len() > int(*c.KCP.Spec.Replicas) {
+		return true
+	}
+	return false
+}
+
+// NeedsScaleUp returns whether the control plane needs to add a machine to the
+// cluster. Scale up can be caused by a rollout, where the cluster will
+// replace outdated machines one by one, or by the more common case of having
+// fewer machines than the number of desired replicas.
+func (c *ControlPlane) NeedsScaleUp() bool {
+	if c.needsReplacementMachine() {
```

Comment:
We may want to have a comment here about not to move this check below to prioritize upgrade over scale up.
Or it may not be a bad idea to move this out of this function and add to the switch logic in the reconcile:
```
case controlPlane. needsReplacementMachine():
   return r.scaleUpControlPlane(ctx, cluster, kcp, controlPlane)
```